server/errors/errors.go

```diff
@@ -0,0 +1,174 @@
+package errors
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/datastore/redis"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	kitlog "github.com/go-kit/kit/log"
+	redigo "github.com/gomodule/redigo/redis"
+	"github.com/mna/redisc"
+	"github.com/rotisserie/eris"
+)
+
+var errCh chan error
+var running int32
+
+type errorHandler struct {
+	pool   fleet.RedisPool
+	logger kitlog.Logger
+}
+
+type ErrorFlusher interface {
+	Flush() ([]string, error)
+}
+
+func NewHandler(ctx context.Context, pool fleet.RedisPool, logger kitlog.Logger) ErrorFlusher {
+	errCh = make(chan error)
+
+	e := &errorHandler{
+		pool:   pool,
+		logger: logger,
+	}
+	go e.handleErrors(ctx)
+	return e
+}
+
+func (e *errorHandler) Flush() ([]string, error) {
+	errorKeys, err := redis.ScanKeys(e.pool, "error:*")
+	if err != nil {
+		return nil, err
+	}
+	keysBySlot := redis.SplitRedisKeysBySlot(e.pool, errorKeys...)
+	var errors []string
+	for _, qkeys := range keysBySlot {
+		gotErrors, err := e.collectBatchErrors(qkeys)
+		errors = append(errors, gotErrors...)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return errors, nil
+}
+
+func (e *errorHandler) collectBatchErrors(errorKeys []string) ([]string, error) {
+	var errors []string
+
+	conn := e.pool.Get()
+	defer conn.Close()
+
+	if rc, err := redisc.RetryConn(conn, 3, 100*time.Millisecond); err == nil {
```

Comment:
You should not use a `RetryConn` here, the goal of the batch is so that it is safe to use `conn.Send` as all keys in the batch hash to the same slot (so are guaranteed to live on the same cluster node). If we expect a lot of keys to process, it's better to use `Send` (and fail in the unlikely event that a failover is going on) rather than follow redirections but make a lot of single `Do` calls.

Reply:
Ah I see that you get it and then delete, so other options to optimize this: you can use `MGET` to get all keys in one go (it's from redis 1.0), and then `DEL` with multiple keys to delete them all in one go too.

Reply:
Ah, I like that!